include/deal.II/grid/connectivity.h

```diff
@@ -0,0 +1,1477 @@
+        AssertDimension(dim, 3);
+
+        return quad_types;
+      }
+
+      inline const std::vector<unsigned int> &
+      entity_types(const unsigned int structdim) const
+      {
+        if (structdim == dim)
+          return cell_types;
+
+        // for vertices/lines the entity types are clear (0/1)
+        AssertDimension(structdim, 2);
+        AssertDimension(dim, 3);
+
+        return quad_types;
+      }
+
+      inline CRS<T> &
+      entity_to_entities(const unsigned int from, const unsigned int to)
+      {
+        if (from == dim && to == dim)
+          return neighbors;
+        else if (from == dim && to == dim - 1)
+          return cell_entities;
+        else if (dim == 3 && from == 2 && to == 0)
+          return quad_vertices;
+        else if (dim == 3 && from == 2 && to == 1)
+          return quad_lines;
+        else if (from == 1 && to == 0)
+          return line_vertices;
+
+        Assert(false, ExcNotImplemented());
+
+        return cell_entities;
+      }
+
+      inline const CRS<T> &
+      entity_to_entities(const unsigned int from, const unsigned int to) const
+      {
+        if (from == dim && to == dim)
+          return neighbors;
+        else if (from == dim && to == dim - 1)
+          return cell_entities;
+        else if (dim == 3 && from == 2 && to == 0)
+          return quad_vertices;
+        else if (dim == 3 && from == 2 && to == 1)
+          return quad_lines;
+        else if (from == 1 && to == 0)
+          return line_vertices;
+
+        Assert(false, ExcNotImplemented());
+
+        return cell_entities;
+      }
+
+    private:
+      const unsigned int        dim;
+      std::vector<unsigned int> cell_types;
+
+      CRS<T> line_vertices;
+
+      std::vector<unsigned char> line_orientation;
+
+      CRS<T> quad_vertices;
+      CRS<T> quad_lines;
+
+      std::vector<unsigned char> quad_orientation;
+
+      CRS<T> cell_entities;
+      CRS<T> neighbors;
+
+      std::vector<unsigned int> quad_types;
+    };
+
+
+
+    /**
+     * Create the transposed of a compressed row storage matrix.
+     */
+    template <typename T>
+    CRS<T>
+    transpose(const CRS<T> &con)
+    {
+      auto        con_copy = con;
+      auto &      col      = con_copy.col;
+      const auto &ptr      = con_copy.ptr;
+
+      CRS<T> con_t;
+      auto & col_t = con_t.col;
+      auto & ptr_t = con_t.ptr;
+
+      for (unsigned int i = 0; i < ptr.size() - 1; i++)
+        std::sort(col.data() + ptr[i], col.data() + ptr[i + 1]);
+
+      col_t.clear();
+      ptr_t.clear();
+
+      col_t.reserve(col.size());
+      ptr_t.reserve(
+        col.size() == 0 ? 0 : *std::max_element(col.begin(), col.end()));
+
+      std::vector<std::pair<T, T>> temp;
+      temp.reserve(col.size());
+
+      for (unsigned int i = 0; i < ptr.size() - 1; i++)
+        for (std::size_t j = ptr[i]; j < ptr[i + 1]; j++)
+          temp.emplace_back(col[j], i);
+
+      std::sort(temp.begin(), temp.end());
```

Comment:
While this is probably one of the shortest ways to transpose a matrix, I am not sure how fast the sorting is compared to all other stuff you do in the setup of the triangulation. If this is an important operation, I believe I could come up with something that should be faster. But let's not optimize it in case we don't need to.

Reply:
Let's postpone this!

Reply:
@kronbichler I have rewritten the function so that it does not need to transpose and gained some additional speedup. Currently, I am at 5.8 (non-compatibility mode).